src/verifier/verifier.rs

```diff
@@ -162,19 +160,40 @@ impl WhammVisitor<Option<DataType>> for TypeChecker<'_> {
         // type check body
 
         self.table.enter_named_scope(&function.name.name);
-        self.visit_block(&function.body);
+        let check_ret_type = self.visit_block(&function.body);
         let _ = self.table.exit_scope();
 
-        // return type
-        todo!();
+        //figure out how to deal with void functions (return type is ())
+        if check_ret_type != Some(DataType::AssumeGood) || check_ret_type != function.return_ty {
+            self.err.type_check_error(
+                false,
+                format! {"Return type of function {} does not match the return type specified in the function signature", function.name.name},
+                &function.name.loc.clone().map(|l| l.line_col),
+            );
+        }
+        //return the type of the fn
+        function.return_ty.clone()
```

Comment:
Just return the `check_ret_type` to avoid an extra `clone()`

Reply:
it has to be cloned because of Option<DataType> causing a "move out"